src/stream/src/executor/sync_kv_log_store.rs

```diff
@@ -0,0 +1,1045 @@
+                }
+                LogStoreBufferItem::Flushed {
+                    start_seq_id,
+                    end_seq_id,
+                    ..
+                } => {
+                    row_count += (end_seq_id - start_seq_id) as usize;
+                }
+                LogStoreBufferItem::Barrier { .. } => {
+                    epoch_count += 1;
+                }
+                LogStoreBufferItem::UpdateVnodes(_) => {}
+            }
+        }
+        self.metrics.buffer_unconsumed_epoch_count.set(epoch_count);
+        self.metrics.buffer_unconsumed_row_count.set(row_count as _);
+        self.metrics
+            .buffer_unconsumed_item_count
+            .set(self.buffer.len() as _);
+        self.metrics.buffer_unconsumed_min_epoch.set(
+            self.buffer
+                .front()
+                .map(|(epoch, _)| *epoch)
+                .unwrap_or_default() as _,
+        );
+    }
+}
+
+impl<S> Execute for SyncedKvLogStoreExecutor<S>
+where
+    S: StateStore,
+{
+    fn execute(self: Box<Self>) -> BoxedMessageStream {
+        self.execute_inner().boxed()
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use pretty_assertions::assert_eq;
+    use risingwave_common::catalog::Field;
+    use risingwave_common::hash::VirtualNode;
+    use risingwave_common::test_prelude::*;
+    use risingwave_common::util::epoch::test_epoch;
+    use risingwave_storage::memory::MemoryStateStore;
+
+    use super::*;
+    use crate::common::log_store_impl::kv_log_store::test_utils::{
+        check_stream_chunk_eq, gen_test_log_store_table, test_payload_schema,
+    };
+    use crate::common::log_store_impl::kv_log_store::KV_LOG_STORE_V2_INFO;
+    use crate::executor::test_utils::MockSource;
+
+    fn init_logger() {
+        let _ = tracing_subscriber::fmt()
+            .with_env_filter(tracing_subscriber::EnvFilter::from_default_env())
+            .with_ansi(false)
+            .try_init();
+    }
+
+    // test read/write buffer
+    #[tokio::test]
+    async fn test_read_write_buffer() {
+        init_logger();
+        let schema = Schema {
+            fields: vec![
+                Field::unnamed(DataType::Int64),
+                Field::unnamed(DataType::Int64),
+            ],
+        };
+        let pk_indices = vec![0];
+        let (mut tx, source) = MockSource::channel();
+        let source = source.into_executor(schema.clone(), pk_indices.clone());
+
+        let vnodes = Some(Arc::new(Bitmap::ones(VirtualNode::COUNT_FOR_TEST)));
+
+        let pk_info = &KV_LOG_STORE_V2_INFO;
+        let table = gen_test_log_store_table(pk_info);
+
+        let log_store_executor = SyncedKvLogStoreExecutor::new(
+            ActorContext::for_test(123),
+            table.id,
+            KvLogStoreReadMetrics::for_test(),
+            KvLogStoreMetrics::for_test(),
+            LogStoreRowSerde::new(&table, vnodes, pk_info),
+            0,
+            MemoryStateStore::new(),
```

Comment:
In the future we'd better add test for recovery and update vnode bitmap, and we may use hummock state store for the test then.